api/include/opentelemetry/trace/trace_state.h

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <cstring>
+
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/nostd/unique_ptr.h"
+
+namespace opentelemetry
+{
+namespace trace
+{
+
+/**
+ * TraceState carries tracing-system specific context in a list of key-value pairs. TraceState
+ * allows different vendors to propagate additional information and inter-operate with their legacy
+ * Id formats.
+ *
+ * Implementation is optimized for a small list of key-value pairs.
+ *
+ * Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter,
+ * and can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
+ * forward slashes /.
+ *
+ * Value is opaque string up to 256 characters, of printable ASCII RFC0020 characters (i.e. the
+ * range 0x20 to 0x7E) except comma , and equals =.
+ */
+class TraceState
+{
+public:
+  static constexpr int kKeyMaxSize       = 256;
+  static constexpr int kValueMaxSize     = 256;
+  static constexpr int kMaxKeyValuePairs = 32;
+
+  // Class to store key-value pairs.
+  class Entry
+  {
+  public:
+    Entry() noexcept = default;
```

Comment:
`Entry` will also need a copy constructor and a destructor (to copy members and to free members).

Reply:
Do we still need a destructor if we use `nostd::unique_ptr<const char[]>`?

Reply:
No, that will be taken care of.